src/sentry/grouping/enhancer/__init__.py

```diff
@@ -463,6 +481,66 @@ def visit_quoted_ident(self, node, children):
         return node.match.groups()[0].lstrip("!")
 
 
+def _merge_values(frames: Sequence[dict[str, Any]], cache_key: str) -> None:
+    """This will merge the cached values if any is found for this stacktrace."""
+    changed_frames_values = cache.get(cache_key)
+    if not changed_frames_values:
+        return
+
+    try:
+        for frame, changed_frame_values in zip(frames, changed_frames_values):
+            frame["in_app"] = changed_frame_values["in_app"]
+            set_path(frame, "data", "category", value=changed_frame_values["category"])
+
+        logger.info("We have merged the cached stacktrace to the incoming one.")
+    except Exception:
+        logger.exception("We have failed to update the stacktrace from the cache.")
+
+
+def _cache_changed_frame_values(frames: Sequence[dict[str, Any]], cache_key: str) -> None:
+    # XXX: How do we make sure that we stay in sync with actions's apply_modifications_to_frame?
+    changed_frames_values = [
+        {
+            "in_app": frame.get("in_app"),  # Based on FlagAction
+            "category": get_path(frame, "data", "category"),  # Based on VarAction's
+        }
+        for frame in frames
+    ]
+    cache.set(cache_key, changed_frames_values)
+
+
+def _generate_matching_frames(
+    frames: Sequence[dict[str, Any]], platform: str
+) -> tuple[list[dict[str, Any]], str]:
+    """Return mock frames which are used for matching rules and a fingerprint representing the stacktrace."""
+    matched_frames = []
+    failed_to_hash = False
+    stacktrace_hash = md5()
+    stacktrace_fingerprint = ""
+    for frame in frames:
+        match_frame = create_match_frame(frame, platform)
+        matched_frames.append(match_frame)
+
+        if not failed_to_hash:
```

Comment:
Why do we stop hashing after one failure?

Reply:
I took sometime going through it in my head and it is safe.

My concern was: Could two different stacktraces be considered the same if one of them had a frame hashing failure but not the other one? The answer after thinking it through is no.

I will address it.